pkg/helm/watcher/controller/helm_watcher.go

```diff
@@ -0,0 +1,140 @@
+package controller
+
+import (
+	"context"
+
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	"github.com/go-logr/logr"
+	"github.com/helm/helm/pkg/chartutil"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/weaveworks/weave-gitops/pkg/helm"
+	"github.com/weaveworks/weave-gitops/pkg/helm/watcher/cache"
+)
+
+const (
+	watcherFinalizer = "finalizers.helm.watcher"
+)
+
+// HelmWatcherReconciler runs the `reconcile` loop for the watcher.
+type HelmWatcherReconciler struct {
+	client.Client
+	Cache       cache.Cache
+	RepoManager helm.HelmRepoManager
+}
+
+// +kubebuilder:rbac:groups=helm.watcher,resources=helmrepositories,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=helm.watcher,resources=helmrepositories/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=helm.watcher,resources=helmrepositories/finalizers,verbs=get;create;update;patch;delete
+
+// Reconcile is either called when there is a new HelmRepository or, when there is an update to a HelmRepository.
+// Because the watcher watches all helmrepositories, it will update data for all of them.
+func (r *HelmWatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContextOrDiscard(ctx)
+
+	// get source object
+	var repository sourcev1.HelmRepository
+	if err := r.Get(ctx, req.NamespacedName, &repository); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Add our finalizer if it does not exist
+	if !controllerutil.ContainsFinalizer(&repository, watcherFinalizer) {
+		patch := client.MergeFrom(repository.DeepCopy())
+		controllerutil.AddFinalizer(&repository, watcherFinalizer)
+
+		if err := r.Patch(ctx, &repository, patch); err != nil {
+			log.Error(err, "unable to register finalizer")
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Examine if the object is under deletion
+	if !repository.ObjectMeta.GetDeletionTimestamp().IsZero() {
+		return r.reconcileDelete(ctx, repository)
+	}
+
+	if repository.Status.Artifact == nil {
+		// This should not occur because the predicate already checks for artifact's existence, but we do this as a
+		// precaution in case that was circumvented.
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("found the repository: ", "name", repository.Name)
```

Comment:
Can I suggest doing one of two things...

Either add a function like:

```go
func namespacedName(o client.Object) types.NamespacedName {
       return types.NamespacedName{
               Name:      o.GetName(),
               Namespace: o.GetNamespace(),
       }
}
```
Which would let you simplify things to:

```go
log.Info("deleting repository cache", "repository", namespacedName(repository))
```

Because `types.NamespacedName` implements `String()`.

OR alternatively...

you could do...

```go
log := logr.FromContextOrDiscard(ctx).WithValues("repository", req.NamespacedName)
```
And then pass that into the other functions

Feel free to create a sublogger from the one if you prefer that...

Reply:
The idea was to be somewhat consistent because not all of the functions are called with a logger context. But yeah I can do that.

Reply:
For the record, what I wanted to say is that `req` is from the watcher, `Put` is the only operation called through the watcher where req is available. For the other two, these will have to be constructed from namespace and name and then passed in from GetProfiles and GetProfileValues where they are called.